tools/autograd/gen_inplace_or_view_type.py

```diff
@@ -333,7 +332,12 @@ def get_creation_meta_in_mode(original: str) -> str:
     return call, rhs_value
 
 def modifies_arguments(f: NativeFunction) -> bool:
-    return f.func.kind() in [SchemaKind.inplace, SchemaKind.out]
+    # See Note [Functionalization: Mutation Removal]
+    # replace_() is a special operator that's only used by the functionalization pass,
+    # and should always call its composite kernel implementation.
+    # The only reason it goes through the dispatcher is to show up in python traces.
```

Comment:
But... should it show up in python traces?

Reply:
I'm ok with taking this out of the PR (and making replace_ not be a NativeFunction), mostly because functionalized tracing for functorch won't fully work by the end of this PR anyway. But my initial reasoning for adding this is below.

---

So, my understanding is that it will eventually be useful for functorch to be able to functionalize programs, trace them and pass off the functionalized trace to an AOT compiler cc @Chillee . Given a program like this:

```
a = ...
b = a.view(...)
b.add_(1)
```

It would be useful if we could feed an AOT a graph like this:
```
x_1 = ...
x_2 = torch.ops.aten.view_copy(x_1)
x_3 = torch.ops.aten.add(x_1, 1)  # inplace becomes out-of-place, magically remember to use x_3 everywhere we used to use x_1
x_4 = torch.ops.aten.add(x_2, 1) same with the view
```

Except we can't actually enforce that "use x_3 everywhere we used to use x_1 from now on" shows up in the graph, just through functionalization machinery. So the next best thing we can do is add breadcrumbs to the traced output that would make it easier to manually fix-up the trace after it's been collected. Something like:
```
x_1 = ...
x_2 = torch.ops.aten.view_copy(x_1)
tmp1 = torch.ops.aten.add(x_1, 1)
x_1 = torch.ops.aten.replace(x_1, tmp1)  # breadcrumb inserted; it should be pretty easy to parse this graph, look for instances of `replace_(a, b)` and manually replace usages of `a` with `b` later in the trace
tmp2 = torch.ops.aten.add(x_2, 1)
x_2 = torch.ops.aten.replace(x_2, tmp2)
```

Reply:
I think you're trying to solve a problem that doesn't exist (except for a very limited special case, which I'll address at the end).

Remember that tracing operates at the very base of the functorch transformation stack. For example, if functionalization was the only pass we were applying to a user program, you'd have FunctionalTensors whizzing around, and their inner Tensors that they are holding are actually the ones we're recording tracing on. You DO NOT trace at all levels of the functorch stack.

This means that your example program should trace to this:

```
a0 = ...
b0 = a0.view_copy(...)
a1 = a0.add(1)
b1 = b0.add(1)
```

No "replacements" necessary; the replaces that you did on FunctionalTensor already did the job. (And even if you did want the replaces in the trace, you wouldn't get them, because you only ever call replace on FunctionalTensor).

The special case where this matters is something like this:

```
def f(x: Tensor):
    x.add_(2)
```

This will trace into `x1 = x0.add(2)` but that's not semantics preserving; at the end of compiling f, the compiler has to insert a bunch of `copy_` to propagate the changes back. Well, but you should have already written this code (it doesn't seem to be in this PR, but I know we've talked about it), and so these should show up in the trace too, all at the very end.

We've had similar variants of this conversation several times in the past; what might be good for developing intuition here is to actually run your functionalization pass on top of functorch tracer and look at some of the traces and understand why they look the way they do.